### PR TITLE
ENH: Use versioneer to manage numpy versions.

### PR DESCRIPTION
**Pull request by charris**
The new tags look like '1.21.0.dev0+98.gaa0453721f', where '98' is the
number of commits since the 1.21.0 branch was started and 'aa0453721f'.
The chosen form may be specified in the 'setup.cfg' file. This PR adds
two new files 'numpy/_version.py' and 'numpy/version.py'. The latter
is kept because it is part of the public API and is actually used by
some downstream projects, but it is no longer dynamically created.

See https://github.com/python-versioneer/python-versioneer/ for more
information.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -422,7 +350,7 @@ def parse_setuppy_commands():
 
 
 def get_docs_url():
```
Comment (charris) [inline setup.py:352]: get_docs_url may need fixing. Not sure how document upload is handled here.

**Comment by charris** [comment]
Not sure why circleCI is failing. Enlightenment is sought.

**Comment by charris** [comment]
The travis failures are probably due to not having the tags, a git handling problem.

**Comment by seberg** [comment]
The circleCI failures look at a lot as if the doctest is running on an older version of NumPy, possibly a 1.19.x?! But I have no enlightenment as to why that would happen...

**Comment by mattip** [comment]
The circleci log for the "build numpy" step has this, so it seems it is building the wheel. Then something, maybe scipy? needs a later numpy than 0.3.0 so it is uninstalling and installing 1.19.4 instead.
``
Building wheels for collected packages: numpy
  Building wheel for numpy (PEP 517) ... 
  Created wheel for numpy: filename=numpy-0.3.0+24119.g25ca33090-cp38-cp38-linux_x86_64.whl size=20680922 sha256=092e1a02b0d038e6df89f48b3aed31dd4af34af526cb7c652ad4d42801e20f6e
  Stored in directory: /tmp/pip-ephem-wheel-cache-dix7tdqn/wheels/ae/2d/57/93f7b3ddf6766cc8c67d5849a172165873ebd8964ed9f8f850
Successfully built numpy
....
...
...
Installing collected packages: numpy
Successfully installed numpy-0.3.0+24119.g25ca33090`
  Attempting uninstall: numpy
    Found existing installation: numpy 0.3.0+24119.g25ca33090
    Uninstalling numpy-0.3.0+24119.g25ca33090:
      Successfully uninstalled numpy-0.3.0+24119.g25ca33090
```

**Comment by mattip** [comment]
If you want this to succeed, I think you should go ahead and push the dev0 tag to master, and then continue with the PR

File: numpy/version.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from ._version import get_versions
+
+__ALL__ = ['version', 'full_version', 'git_revision', 'release']
+
+vinfo = get_versions()
+version = vinfo["version"]
+full_version = vinfo['version']
+git_revision = vinfo['full-revisionid']
```
Comment (BvB93) [inline numpy/version.py:8]:
```suggestion
version: str = vinfo["version"]
full_version: str = vinfo['version']
git_revision: str = vinfo['full-revisionid']
```
Would it be possible to add a few annotations here? This should be enough to annotate the entire module.
In principle we could also reintroduce `version.pyi`, but I feel this is change would be quite harmless.

**Comment by charris** [comment]
> I think you should go ahead and push the dev0 tag to master

Did it a while ago, tagged `v1.21.0dev0`. Note that you need to pull tags separately to update your branch, `git fetch upstream --tags`. That is a bit unfortunate as it is just one thing more to go wrong :)

**Comment by charris** [comment]
Or maybe a pull just pulls down the tags that can be reached in the pulled branch. I've always found that a bit confusing :) You can see the most recent tag in  a branch with `git describe`

File: numpy/version.py (hunk 2 above)
Comment (charris) [reply numpy/version.py:8]: Sure. How do I do that?

Comment (charris) [reply numpy/version.py:8]: Oh, NVM, I see. I wondered why the original was done that way :)

Comment (charris) [reply numpy/version.py:8]: Done. I'll wait before pushing in case more fixes are suggested.

**Comment by charris** [comment]
The CircleCI version -- 0.3.0 -- is weird. Should be `1.21.0dev` with some extras.

File: setup.py [hunk 3]
```diff
@@ -24,13 +24,48 @@
...
+# BEFORE importing setuptools, remove MANIFEST. Otherwise it may not be
+# properly updated when the contents of directories change (true for distutils,
+# not sure about setuptools).
+if os.path.exists('MANIFEST'):
+    os.remove('MANIFEST')
 
+# We need to import setuptools here in order for it to persist in sys.modules.
+# Its presence/absence is used in subclassing setup in numpy/distutils/core.py,
+# which may not be the most robust design.
+import setuptools
```
Comment (charris) [inline setup.py:58]: This was tricky to discover, we have been doing it wrong all along.

**Comment by charris** [comment]
> Did it a while ago, tagged v1.21.0dev0.

And now I checked, and it wasn't there. I thought I pushed it... It is there now.

File: setup.py (hunk 3 above)
Comment (eric-wieser) [reply setup.py:63]: I think the rationale is that setuptools has to be imported before distutils; therefore, if `setuptools` is not already in `sys.modules` before numpy.distutils is imported, it's too late. I thought we already handled this, as I remember warnings being emitted about it...

File: setup.py [hunk 4]
```diff
@@ -24,13 +24,47 @@
...
+# numpy __init__ can detect if it is being loaded by the setup routine, to
+# avoid attempting to load components that aren't built yet.  While ugly, it's
+# a lot more robust than what was previously being used.
+builtins.__NUMPY_SETUP__ = True
+
+# Needed for backwards code compatibility below and in some CI scripts.
+FULLVERSION = versioneer.get_version()
+ISRELEASED = 'dev' not in FULLVERSION
+MAJOR, MINOR, MICRO = FULLVERSION.split('.')[:3]
+VERSION = f"{MAJOR}.{MINOR}.{MICRO}"
```
Comment (eric-wieser) [inline setup.py:40]: We can't use `f-strings` in setup.py, it makes users of Python 3.5 unable to use `pip` to install old numpy versions.

Comment (charris) [reply setup.py]: Fixed.

**Comment by charris** [comment]
The new procedure required for branching/releasing needs to be documented. That will be a followup PR if this goes in to make discussion easier.

File: setup.py (hunk 3 above)
Comment (charris) [reply setup.py:65]: It caused a problem with sdist that was subclassed before the import. The root problem may have been due to initializing the `cmdclass` with `versioneer.get_cmdclass` that took as its argument the cmdclass from `numpy/distutils/core.py`.  We will see if any other problems show up.

**Comment by mattip** [comment]
Is it typical to vendor `versioneer.py` into the root of the repo rather than to add it to the `pyproject.toml` build requirements? If we want to vendor it, maybe we can move it to `tools`?

**Comment by charris** [comment]
> Is it typical to vendor versioneer.py into the root of the repo

Yes, versioneer vendors itself into the root directory. See also matplotlib and pandas. The install instructions are
``` 
- pip install versioneer to somewhere in your $PATH
- add a [versioneer] section to your setup.cfg (see Install)
- run versioneer install in your source tree, commit the results
- Verify version information with python setup.py version
```

**Review by mattip — APPROVED** [review]
LGTM. I will merge it soon if there are no more comments

**Comment by mattip** [comment]
Thanks @charris

**Comment by charris** [comment]
Thanks Matti. Now to test it on numpy-wheels...

**Comment by charris** [comment]
Works on nightly. The version does change if something is merged to master between wheel builds because the number of development commits is part of  the version, but that doesn't seem too bad, especially as the nightly builds take place at a time when few PRs seem to get merged.

**Comment by mattip** [comment]
If it becomes a problem we could think about various mitigation strategies to tie the commit we build to something like time-of-the-cron trigger or so.

**Comment by charris** [comment]
Strictly speaking, they _are_ different  versions because they are built with different master branches.

**Comment by charris** [comment]
Hmm, all the wheels have `dirty` in the version because there are uncommitted changes due to patching in the the licenses and such. I don't see a way to disable that via a setting.

**Pull request merged**